snuba/datasets/errors_replacer.py

```diff
@@ -206,93 +213,129 @@ def set_project_needs_final(
     project_id: int,
     state_name: Optional[ReplacerState],
     replacement_type: ReplacementType,
-) -> Optional[bool]:
-    key, type_key = get_project_needs_final_key_and_type_key(project_id, state_name)
+) -> None:
+    key, type_key = build_project_needs_final_key_and_type_key(project_id, state_name)
     p = redis_client.pipeline()
-    p.set(key, True, ex=settings.REPLACER_KEY_TTL)
+    p.set(key, time.time(), ex=settings.REPLACER_KEY_TTL)
     p.set(type_key, replacement_type, ex=settings.REPLACER_KEY_TTL)
-    [final_was_set, _] = p.execute()
-    return bool(final_was_set)
+    p.execute()
+
+
+@dataclass
+class ProjectsQueryFlags:
+    needs_final: bool
+    group_ids_to_exclude: Sequence[int]
+    replacement_types: Set[str]
+    latest_replacement_time: Optional[datetime]
 
 
 def get_projects_query_flags(
     project_ids: Sequence[int], state_name: Optional[ReplacerState]
-) -> Tuple[bool, Sequence[int], Set[str]]:
-    """\
+) -> ProjectsQueryFlags:
+    """
     1. Fetch `needs_final` for each Project
     2. Fetch groups to exclude for each Project
-    3. Trim groups to exclude ZSET for each Project
+    3. Clean up the old replacement types and groups
     4. Fetch replacement types for each Project
+    5. Fetch latest exclude groups replacement type for each Project
 
-    Returns (needs_final, group_ids_to_exclude, replacement_types)
+    Returns (
+        needs_final,
+        group_ids_to_exclude,
+        replacement_types,
+        latest_replacement_time
+    )
     """
 
     s_project_ids = set(project_ids)
     now = time.time()
     p = redis_client.pipeline()
 
     needs_final_keys_and_type_keys = [
-        get_project_needs_final_key_and_type_key(project_id, state_name)
+        build_project_needs_final_key_and_type_key(project_id, state_name)
         for project_id in s_project_ids
     ]
 
     for needs_final_key, _ in needs_final_keys_and_type_keys:
         p.get(needs_final_key)
 
     exclude_groups_keys_and_types = [
-        get_project_exclude_groups_key_and_type_key(project_id, state_name)
+        build_project_exclude_groups_key_and_type_key(project_id, state_name)
         for project_id in s_project_ids
     ]
-
     for exclude_groups_key, _ in exclude_groups_keys_and_types:
+        # remove stale excluded groups
```

Comment:
Nit: some of these comments of seem unnecessary.

Reply:
to echo what Evan said, commenting to explain what the code does generally means your code is not readable enough. Fixing the code is the preferred option instead of adding comments. So if you feel you needed to explain that, probably refactoring is needed. Here I doubt that comment is needed entirely.

Commenting on the "why" the code does something is different and can be useful a lot. Basically the idea is that the comment provides context, it does not explains what we can already read in the code.

In this specific case, I think you are explaining everything in the docstring. No need to pollute the code, if you want to add more context, please feel free to beef up the docstring

Reply:
I think this comment is better here than in the docstring as it actually tells me what it's doing when it's doing it. I also think that comment is actually communicating intent. Consider this line:

```python3
p.zremrangebyscore(
    exclude_groups_key, float("-inf"), now - settings.REPLACER_KEY_TTL
)
``` 

In order to understand what it's doing you need to:
* Know what `remrangebyscore` does
* its arguments and their order
* how this sorted set is scored

I understand that somebody with a lot of context in the codebase can see this pattern clearly but processing the above without the comment is much harder than with.